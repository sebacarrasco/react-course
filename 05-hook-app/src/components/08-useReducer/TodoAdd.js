import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ( { handleAddTodo } ) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: ""
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        // Validación del input
        if ( description.trim().length <= 1)
        {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        // Se le manda el dispatch con la acción y él ya sabe qué
        // reducer usar y hará render de lo que cambie luego de actualizar el state
        handleAddTodo(newTodo)
        reset();
    };

    return (
        <>
            <h4> Add TODO </h4>
            <hr/>

            <form onSubmit={ handleSubmit }>

                <input
                className="form-control"
                type="text"
                name="description"
                placeholder="Do something..."
                autoComplete="off"
                onChange={ handleInputChange }
                value={ description }
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 w-100"
                >
                    Add
                </button>

            </form>
        </>
    )
}
