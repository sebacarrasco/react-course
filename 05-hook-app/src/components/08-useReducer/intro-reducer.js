

const initialState = [{
    id: 1,
    todo: "primera tarea",
    done: false
}];

const todoReducer = (state=initialState, action) => {

    // El  action?.type quiere decir que si action tiene algún valor,
    // entonces le revisa la propiedad type, en caso contrario no se hace nada
    if ( action?.type === "agregar")
    {
        return [...state, action.payload];
    }

    return state;
};

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: "segunda tarea",
    done: false
}

const agregarTodoAction = {
    type: "agregar",
    payload: newTodo // Es solo convención los nombres de type y payload
}

todos = todoReducer(todos, agregarTodoAction);

console.log(todos)