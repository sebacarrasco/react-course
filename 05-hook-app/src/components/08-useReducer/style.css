body{
    padding: 5%
}

.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
}

p{
    cursor: pointer;
    margin: 0px;
}

.complete {
    text-decoration: line-through;
}