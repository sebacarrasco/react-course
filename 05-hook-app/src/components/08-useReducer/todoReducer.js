
export const todoReducer = ( state=[], action ) => {

    switch (action.type) {
        case "add":
            // Acá se manda el objeto
            return [...state, action.payload];
    
        case "delete":
            // acá se manda el id
            return state.filter(task => task.id !== action.payload);

        case "toggle":
            // acá se manda el id

            // Versión larga:
            // return state.map(task => {
            //     if (task.id === action.payload)
            //     {
            //         return {
            //             ...task,
            //             done: !task.done
            //         };
            //     }
            //     else
            //     {
            //         return task;
            //     }
            // });

            // Versión corta
            return state.map( task => 
                ( task.id === action.payload)
                    ?  {...task, done: !task.done }
                    : task
            )
        default:
            return state;
    }

};