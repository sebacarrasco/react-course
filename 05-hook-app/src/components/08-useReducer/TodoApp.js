import React, { useReducer, useEffect } from 'react';
// import { useForm } from '../../hooks/useForm';
import "./style.css";
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';

const init = () => {
    return JSON.parse(localStorage.getItem("todos")) || [];
}

export const TodoApp = () => {

    // useReducer es una alternativa al useState que se suele usar cuando hay mucha
    // lógica detrás de la modificación del state.
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos))
    }, [todos]);

    const handleDelete = (todoId) => {
        const action = {
            type: "delete",
            payload: todoId
        }

        dispatch(action);
    };

    const handleToggle = (todoId) => {
        dispatch({
            type: "toggle",
            payload: todoId
        });
    };

    const handleAddTodo = ( newTodo ) => {
        dispatch({
            type: "add",
            payload: newTodo
        })
    }

    return (
        <div>
            <h1>TODO App { todos.length } </h1>
            <hr/>

            <div className="row">

                <div className="col-7">
                    <TodoList todos={todos} handleDelete={ handleDelete } handleToggle={ handleToggle }/>
                </div>

            <div className="col-5">
                <TodoAdd handleAddTodo={ handleAddTodo }/>
            </div>

            </div>
        </div>
    )
}
