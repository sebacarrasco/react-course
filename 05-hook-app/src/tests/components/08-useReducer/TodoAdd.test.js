import { shallow } from 'enzyme';
import React from 'react';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';

describe('Tests for <TodoAdd />', () => {

    const handleAddTodo = jest.fn();
    const wrapper = shallow(<TodoAdd handleAddTodo={ handleAddTodo }/>);
    const value = 'A task';

    test('should match the snapshot', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('should not call handleAddTodo', () => {
       
        // wrapper.find("form").simulate("submit", { preventDefault: () => {} });
        // Otra forma para simular el submit:
        const formSubmit = wrapper.find("form").prop("onSubmit"); // Esto retorna una funciÃ³n
        formSubmit({ preventDefault: () => {} });

        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    });
    
    test('should call the function handleAddTodo', () => {

        const input = wrapper.find("input");
        input.simulate('change', { target: { value, name: 'description' } });
        wrapper.find("form").simulate("submit", { preventDefault: () => {} });
        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith({ desc: value, done: false , id: expect.any(Number)});
        expect(input.prop("value")).toBe("");
        
    });
    
    

    
});
