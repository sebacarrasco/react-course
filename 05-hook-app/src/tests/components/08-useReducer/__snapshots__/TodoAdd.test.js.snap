// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <TodoAdd /> should match the snapshot 1`] = `
<Fragment>
  <h4>
     Add TODO 
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Do something..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 w-100"
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
